cns/deviceplugin/pluginmanager.go

```diff
@@ -0,0 +1,147 @@
+	}
+}
+
+// PluginManager runs device plugins for vnet nics and ib nics
+type PluginManager struct {
+	Logger        *zap.Logger
+	VnetNICPlugin *Plugin
+	IBNICPlugin   *Plugin
+	options       pluginManagerOptions
+}
+
+func NewPluginManager(l *zap.Logger, initialVnetNICCount, initialIBNICCount int, opts ...pluginManagerOption) *PluginManager {
+	logger := l.With(zap.String("component", "devicePlugin"))
+	socketWatcher := NewSocketWatcher(logger)
+	options := pluginManagerOptions{
+		devicePluginDirectory: defaultDevicePluginDirectory,
+		kubeletSocket:         v1beta1.KubeletSocket,
+		deviceCheckInterval:   defaultDeviceCheckInterval,
+	}
+	for _, o := range opts {
+		o(&options)
+	}
+	return &PluginManager{
+		Logger: logger,
+		VnetNICPlugin: NewPlugin(logger, string(v1alpha1.DeviceTypeVnetNIC), socketWatcher,
+			getSocketName(options.devicePluginDirectory, v1alpha1.DeviceTypeVnetNIC), initialVnetNICCount,
+			options.kubeletSocket, options.deviceCheckInterval),
+		IBNICPlugin: NewPlugin(logger, string(v1alpha1.DeviceTypeInfiniBandNIC), socketWatcher,
+			getSocketName(options.devicePluginDirectory, v1alpha1.DeviceTypeInfiniBandNIC), initialIBNICCount,
+			options.kubeletSocket, options.deviceCheckInterval),
+		options: options,
+	}
+}
+
+// Run runs the plugin manager until the context is cancelled or error encountered
+func (p *PluginManager) Run(ctx context.Context) error {
+	// clean up any leftover state from previous failed plugins
+	// this can happen if the process crashes before it is able to clean up after itself
+	if err := p.cleanOldState(); err != nil {
+		return errors.Wrap(err, "error cleaning state from previous plugin process")
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(defaultServerCountForDevicePlugin) // starting two goroutines to wait on
+	go func() {
+		defer wg.Done()
+		p.VnetNICPlugin.Run(ctx)
+	}()
+
+	go func() {
+		defer wg.Done()
+		p.IBNICPlugin.Run(ctx)
+	}()
+
+	wg.Wait()
+	return nil
+}
+
+func (p *PluginManager) TrackDevices(deviceType v1alpha1.DeviceType, count int) error {
+	switch deviceType {
+	case v1alpha1.DeviceTypeInfiniBandNIC:
+		p.IBNICPlugin.UpdateDeviceCount(count)
+	case v1alpha1.DeviceTypeVnetNIC:
+		p.VnetNICPlugin.UpdateDeviceCount(count)
+	default:
+		return ErrUnknownDeviceType
+	}
+
+	return nil
+}
+
+func (p *PluginManager) cleanOldState() error {
+	entries, err := os.ReadDir(p.options.devicePluginDirectory)
+	if err != nil {
+		return errors.Wrap(err, "error listing existing device plugin sockets")
+	}
+	for _, entry := range entries {
+		if strings.HasPrefix(entry.Name(), path.Base(getSocketPrefix(p.options.devicePluginDirectory, v1alpha1.DeviceTypeVnetNIC))) ||
+			strings.HasPrefix(entry.Name(), path.Base(getSocketPrefix(p.options.devicePluginDirectory, v1alpha1.DeviceTypeInfiniBandNIC))) {
+			// try to delete it
+			f := path.Join(p.options.devicePluginDirectory, entry.Name())
+			if err := os.Remove(f); err != nil {
+				return errors.Wrapf(err, "error removing old socket %s", f)
```

Comment:
I think implicitly quoting paths might be nicer in this circumstance:

```suggestion
				return errors.Wrapf(err, "error removing old socket %q", f)
```